Fixed #30311 -- Restored ability to override global admin actions.

hashlash opened the pull request:
https://code.djangoproject.com/ticket/30311

hashlash commented:
Actually I have an alternative implementation which will make sure the resulting actions are unique

```py
def _get_base_actions(self):
    """Return the list of actions, prior to any request-based filtering."""
    actions = {}

    # Gather actions from the admin site first
    for (name, func) in self.admin_site.actions:
        description = getattr(func, 'short_description', name.replace('_', ' '))
        actions[name] = (func, name, description)
    # Add actions from this ModelAdmin.
    for action in self.actions or []:
        actions[name] = (func, name, description)
    # get_action might have returned None, so filter any of those out.
    return filter(None, actions.values())
```

But that will bypass the `admin.E130` system check and make the `modeladmin.test_checks.ActionsCheckTests.test_actions_not_unique` test fail.

Any thoughts on this?

charettes commented on django/contrib/admin/options.py:
```diff
@@ -850,9 +850,12 @@ def action_checkbox(self, obj):
     def _get_base_actions(self):
         """Return the list of actions, prior to any request-based filtering."""
         actions = []
+        local_action_names = [self.get_action(action)[1] for action in self.actions or []]
```
Use a `set` if this value is only used for containment checks

```suggestion
        local_action_names = {self.get_action(action)[1] for action in self.actions} if self.actions else set()
```

charettes replied:
I don't have a strong feeling about either, this approach avoids an unnecessary intermediary list creation though.

felixxm commented:
@carltongibson IMO we don't need a release notes, it was possible to overwrite site-wide admin actions actions before Django 2.2, so it's not really a new feature :thinking:  We're fixing a small regression.

carltongibson commented:
OK. Agreed. 👍

felixxm commented on django/contrib/admin/options.py:
```diff
@@ -850,9 +850,12 @@ def action_checkbox(self, obj):
     def _get_base_actions(self):
         """Return the list of actions, prior to any request-based filtering."""
         actions = []
+        admin_action_names = {self.get_action(action)[1] for action in self.actions or []}
```
`self.actions` already contains action names so we don't need to wrap them in `get_action()`. Moreover for a nonexistent name it raises `TypeError: 'NoneType' object is not subscriptable`. We've already ignored them in `filter(None, actions)` so it's safe to use `self.actions`.

I will adjust this.

hashlash replied:
But wouldn't it failed if the `ModelAdmin`-specific action is defined by passing the function itself (instead of the function name), like the example written here: https://docs.djangoproject.com/en/3.0/ref/contrib/admin/actions/#adding-actions-to-the-modeladmin. I will write an additional test for this.

felixxm replied:
Good point.

felixxm commented:
@hashlash Thanks :+1: I pushed small edits.

hashlash commented:
Ah sorry, I work on the docs and pushed without checking @felixxm comment. Should I remove the docs?

hashlash commented:
I've removed the change on the release note, but I think it's worth mentioning how to override the site-wide admin action in the docs.

And I've added a test regarding [this review](https://github.com/django/django/pull/12591#discussion_r410145697)

felixxm commented:
@hashlash You removed all my edits :confused:

hashlash commented:
Ahh... I'm so sorry I haven't put your edit on the test back :sob:. I'll put it back right away. For your edit on implementation, would you mind check my [comment](https://github.com/django/django/pull/12591#discussion_r410256557) on your review?

Sorryy :(

felixxm commented:
@hashlash np, I pushed them again.

hashlash commented:
Thanks a lot! @felixxm

hashlash commented on django/contrib/admin/options.py:
```diff
@@ -858,9 +858,13 @@ def action_checkbox(self, obj):
     def _get_base_actions(self):
         """Return the list of actions, prior to any request-based filtering."""
         actions = []
+        base_actions = [self.get_action(action) for action in self.actions or []]
```
@charettes mentioned in [his review](https://github.com/django/django/pull/12591#discussion_r395680114) to avoid an unnecessary intermediary list creation. Would it be a problem?

hashlash replied:
Actually, I prefer this implementation for readability

felixxm replied:
We can use generator.

hashlash replied:
Like this?

```py
base_actions = (self.get_action(action) for action in self.actions or [])
```

hashlash commented on django/contrib/admin/options.py:
```diff
@@ -858,9 +858,13 @@ def action_checkbox(self, obj):
         """Return the list of actions, prior to any request-based filtering."""
         actions = []
+        base_actions = [self.get_action(action) for action in self.actions or []]
+        base_action_names = {action[1] for action in base_actions if action}
```
How about:

```py
base_action_names = {name for _, name, _ in filter(None, base_actions)}
```

I think it's more readable because we'll know that `base_actions` is a list of 3-tuple which the second element is the name (and the other two is ignored).

felixxm commented:
@hashlash I decided to reuse `base_actions` and remove `filter()`. In that way we don't iterate `self.actions` twice.

hashlash commented:
Nice approach @felixxm :clap: 

Btw, I'm a bit curious. I was going to propose to directly assign the actions to `actions` list

```py
    def _get_base_actions(self):
        """Return the list of actions, prior to any request-based filtering."""
        actions = (self.get_action(action) for action in self.actions or [])
        # get_action might have returned None, so filter any of those out.
        actions = [action for action in actions if action]
        action_names = {name for _, name, _ in actions}

        # Gather actions from the admin site first
        for (name, func) in self.admin_site.actions:
            if name in action_names:
                continue
            description = getattr(func, 'short_description', name.replace('_', ' '))
            actions.append((func, name, description))

        return actions
```

But it fails some tests (`test_get_actions_respects_permissions` and `test_actions_inheritance`) because the order is reversed: expected `['delete_selected', 'custom_action']` got `['custom_action', 'delete_selected']`.

Does/should `ModelAdmin` maintain the order as site-wide actions first then `ModelAdmin`-specific actions? I can't find any info about the ordering on [the docs](https://docs.djangoproject.com/en/3.0/ref/contrib/admin/actions/).

felixxm commented:
IMO we should keep the current ordering.
